Add option to load historic operators in IR when the operator is deprecated

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #71148

Differential Revision: [D33521300](https://our.internmc.facebook.com/intern/diff/D33521300)

tugsbayasgalan committed:
Add option to load historic operators in IR when the operator is deprecated

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33521300).

gmagogsfm commented on test/cpp/jit/test_upgrader_utils.cpp:
```diff
@@ -58,5 +60,31 @@ TEST(UpgraderUtils, FindIfOpIsCurrent) {
+  std::vector<std::string> schemas = {"foo.bar()", "foo.foo()"};
+
+  // symbol based look up
+  test_only_add_entry("foo.first", dummy_entry[0]);
```
I would suggest using a more exotic name like "an_op_that_does_not_exit" to avoid accidental collisons.

tugsbayasgalan replied:
done

gmagogsfm commented on torch/csrc/jit/frontend/schema_matching.cpp:
```diff
@@ -678,6 +680,17 @@ Value* emitBuiltinCall(
 
+  // we might have seen old historic
+  // ops that are deprecated
+  if (variants.size() == 0) {
```
`variants.empty()`

tugsbayasgalan replied:
done

gmagogsfm commented on torch/csrc/jit/frontend/schema_matching.cpp:
```diff
@@ -678,6 +680,17 @@ Value* emitBuiltinCall(
+  if (variants.size() == 0) {
+    auto oldSchemas =
+        loadPossibleHistoricOps(name.toQualString(), graph_version);
+    for (const auto& old_schema_entry : oldSchemas) {
```
Nit:
reserve space for `upgrader_schemas`

tugsbayasgalan replied:
done

gmagogsfm commented on torch/csrc/jit/ir/ir.h:
```diff
@@ -338,6 +338,8 @@ struct TORCH_API Node {
   topo_position_t topo_position_ = 0;
   // a managing wrapper for Python to allow invalidation
   std::shared_ptr<Wrap<Node>> wrap_;
+  // stores the full schema name, if the operator is historic
```
Consider defining historic more clearly

tugsbayasgalan replied:
done

gmagogsfm commented on torch/csrc/jit/ir/ir.h:
```diff
@@ -362,6 +364,14 @@ struct TORCH_API Node {
     return wrap_;
   }
 
+  c10::optional<std::string> getHistoricSchemaName() {
```
Should this method be const?

tugsbayasgalan replied:
done

gmagogsfm commented on torch/csrc/jit/operator_upgraders/utils.cpp:
```diff
@@ -49,5 +49,27 @@ bool isOpSymbolCurrent(const std::string& name, size_t current_version) {
+
+  for (const auto& entry : get_operator_version_map()) {
+    auto old_symbol_name = entry.first;
+    if (old_symbol_name.find(name) == 0) {
```
`old_symbol_name` is a string I think, and `name` should be an exact match for it, why using `find` which checks if they are substring relationships? Is it to get rid of overload name? If so, we should have more strict logic that separates `old_symbolc_name` into two parts by `.`. Current implementation could be confused by symbol names that prefix each other.

tugsbayasgalan replied:
This is because name is actually a symbol so it doesn't have the overload in it. schema_matching is responsible for finding all the overloads for this name and do schema matching to find the correct overload. But, operator_version_map stores the schema name with its' overload, so we need to check if the old_symbol_name is a valid overload for name. That being said, you are right that the check needs to be stricter.

gmagogsfm commented on torch/csrc/jit/operator_upgraders/utils.h:
```diff
@@ -30,5 +30,9 @@ TORCH_API bool isOpSymbolCurrent(
     const std::string& name,
     size_t current_version);
 
+TORCH_API std::vector<std::string> loadPossibleHistoricOps(
```
Consider adding comment on what this method does.

tugsbayasgalan replied:
done

gmagogsfm commented on torch/csrc/jit/passes/replacement_of_old_operators.cpp:
```diff
@@ -28,20 +28,28 @@ struct OldOpsReplacerWithUpgraders {
+        schema_name = node->getHistoricSchemaName();
+      }
+
+      if (schema_name.has_value()) {
```
Shouldn't it error out when this condition is not met? In fact, the logic would be clearer and less error prone if we check for `!schema_name.has_value()` and error out when true.

tugsbayasgalan replied:
I think it shouldn't error out because nodes don't need to have schemas

tugsbayasgalan committed:
Update on "Add option to load historic operators in IR when the operator is deprecated"


Differential Revision: [D33521300](https://our.internmc.facebook.com/intern/diff/D33521300)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33521300).

tugsbayasgalan committed:
Update on "Add option to load historic operators in IR when the operator is deprecated"


Differential Revision: [D33521300](https://our.internmc.facebook.com/intern/diff/D33521300)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33521300).
